roofit/batchcompute/src/RooBatchCompute.cu

```diff
@@ -0,0 +1,174 @@
+// RooBatchCompute library created September 2020 by Emmanouil Michalainas
+/**
+\file RooBatchCompute.cu
+\class RbcClass
+\ingroup Roobatchcompute
+
+This file contains the code for cuda computations using the RooBatchCompute library.
+**/ 
+#include "rbc.h"
+#include "Batches.h"
+
+#include "TError.h"
+
+#include <algorithm>
+#include <thrust/reduce.h>
+
+#ifndef RF_ARCH
+#error "RF_ARCH should always be defined"
+#endif
+
+namespace RooBatchCompute {
+namespace RF_ARCH {
+
+// For turning preprocessor definition into string
+#define _STRINGIFY(X) #X
+#define STRINGIFY(X) _STRINGIFY(X)
+  
+std::vector<void(*)(Batches)> getFunctions();
+/// This class overrides some RbcInterface functions, for the purpose of providing
+/// a cuda specific implementation of the library.
+class RbcClass : public RbcInterface {
+  private:
+    const std::vector<void(*)(Batches)> _computeFunctions;
+  public:
+    RbcClass() : _computeFunctions(getFunctions())
+    {
+      dispatchCUDA = this; // Set the dispatch pointer to this instance of the library upon loading
+    }
+
+    Architecture architecture() const override { return Architecture::RF_ARCH; };
+    std::string architectureName() const override {
+        // transform to lower case to match the original architecture name passed to the compiler
+        std::string out = STRINGIFY(RF_ARCH);
+        std::transform(out.begin(), out.end(), out.begin(), [](unsigned char c){ return std::tolower(c); });;
+        return out;
+    };
+
+    /** Initialize the cuda computation library.
+    This method needs to be called after the dynamic loading of the cuda instance of the
+    RooBatchCompute library. If cuda is not working properly, it will set the dispatchCUDA
+    pointer to nullptr. **/
+    void init()
+    {
+      cudaError_t err = cudaSetDevice(0);
+      if (err==cudaSuccess) cudaFree(nullptr);
+      else
+      {
+        dispatchCUDA = nullptr;
+        Error( (std::string(__func__)+"(), "+__FILE__+":"+std::to_string(__LINE__)).c_str(), "%s", cudaGetErrorString(err) );
```

Comment:
Is this ROOT's TError.h Error? That does the source file location stuff already.
```suggestion
        Error("RbcClass::init()", cudaGetErrorString(err) );
```

Reply:
Ok!